File: 구현/주문처리시스템/jeffrey/Store/Product.swift
//
//  Product.swift
//  ShoppingMall
//
//  Created by Mephrine on 1/8/25.
//

import Foundation

// MARK: - 스토어 모듈
public struct Product {
    public enum DeliveryCharge {
        case free
        case normal
        case quick
        
        public var cost: Decimal {
            switch self {
            case .free: return 0
            case .normal: return 3_000
            case .quick: return 5_000
            }
        }
    }
    public let name: String
    public let price: Decimal
    /// 수량은 주문 내역 등에서 알 필요가 없기 때문에 internal로 설정.
    let quantity: Int
    let deliveryCharge: DeliveryCharge

    public init(name: String,
                price: Decimal,
                quantity: Int,
                deliveryCharge: DeliveryCharge = .free) {
        self.name = name
        self.price = price
        self.quantity = quantity
        self.deliveryCharge = deliveryCharge
    }
    
    public var hasStock: Bool {
        self.quantity > 0
    }
}

File: 구현/주문처리시스템/julie/Model/Product.kt
data class Product(
    val name: String,
    val price: Int,
    val quantity: Int,
    val deliveryCharge: Int // 상품별 배송비 (0, 3000, 5000)
) {
    init {
        // 상품 수량이 0 이하일 경우 예외를 발생시킴
        if (quantity <= 0) throw CustomError.InvalidQuantity
    }
}
